### PR TITLE
Fix memory related errors

### PR DESCRIPTION
**Pull request by RMTT**
## Description
Fix memory-related errors when using sql to select array type columns or large size data.

## Tests
Create a table with an array type column, and select it in plcontainer

## User Story or Issue
https://www.pivotaltracker.com/story/show/179879954

### CODE REVIEW

File: src/sqlhandler.c [hunk 1]
```diff
@@ -63,7 +63,14 @@ static plcMsgResult *create_sql_result(bool isSelect) {
 	result->types = palloc(result->cols * sizeof(*result->types));
 	result->names = palloc(result->cols * sizeof(*result->names));
 	result->exception_callback = NULL;
+
+        /*
```
Comment (beeender) [inline src/sqlhandler.c:67]: indention is wrong

File: src/sqlhandler.c [hunk 2]
```diff
@@ -310,6 +312,11 @@ plcMessage *handle_sql_message(plcMsgSQL *msg, plcConn *conn, plcProcInfo *pinfo
 						}
 					}
 
+                                        /*
```
Comment (beeender) [inline src/sqlhandler.c:315]: indention is wrong

File: src/sqlhandler.c [hunk 3]
```diff
@@ -63,7 +63,14 @@ static plcMsgResult *create_sql_result(bool isSelect) {
...
+        /*
+         * We do not need free this array because the array iterator will reference it.
+         * This function is under SPI_proc Memory Context, when SPI_finish is invoked, 
+         * the array will be freed automatically. 
+         */
 	resTypes = palloc(result->cols * sizeof(plcTypeInfo));
+
 	for (j = 0; j < result->cols; j++) {
 		fill_type_info(NULL, SPI_tuptable->tupdesc->attrs[j]->atttypid, &resTypes[j]);
 		copy_type_info(&result->types[j], &resTypes[j]);
```
Comment (beeender) [inline src/sqlhandler.c:76]: If the type is not going to be freed, do we still need to copy them here?

**Review by xuebinsu — APPROVED** [review]
+1, better to keep indention consistent.

File: src/sqlhandler.c [hunk 4]
```diff
@@ -63,7 +63,14 @@ static plcMsgResult *create_sql_result(bool isSelect) {
 	result->types = palloc(result->cols * sizeof(*result->types));
 	result->names = palloc(result->cols * sizeof(*result->names));
 	result->exception_callback = NULL;
+
+        /*
+         * We do not need free this array because the array iterator will reference it.
+         * This function is under SPI_proc Memory Context, when SPI_finish is invoked, 
+         * the array will be freed automatically. 
+         */
```
Comment (beeender) [inline src/sqlhandler.c:71]: I am thinking of another possible fix.
The `plcMsgResult` is supposed to be freed at `plcontainer.c:565` the `free_result()`.
Would it be better if we has `plcTypeInfo` as a member in the `plcMsgResult` and free it together in the `free_result()`? Then we have a completed life-control of the type. What do you think?

Comment (beeender) [reply src/sqlhandler.c:71]: One more thing, do we have a test case to cover the array types here?

File: src/sqlhandler.c (hunk 3 above)
Comment (RMTT) [reply src/sqlhandler.c:76]: Yeah, because the `result->types` has a different structure and layout from `resTypes`.

Comment (beeender) [reply src/sqlhandler.c:76]: Would it make sense if we just have `plcTypeInfo` in the `plcMsgResult`

```
typedef struct plcMsgResult {
...
	// Remove this
	// plcType *types;
	// Add
	plcTypeInfo *types;
...
} plcMsgResult;
```

So we don't have to copy, and the types can be freed in `free_result()`?

File: src/sqlhandler.c (hunk 1 above)
Comment (beeender) [reply src/sqlhandler.c:67]: @RMTT Just install this one
https://github.com/editorconfig/editorconfig-vim

**Comment by beeender** [comment]
The minimal reproduce steps for the original issue:

```sql
DROP TABLE IF EXISTS array_test_table;
CREATE TABLE array_test_table(
    content text[]
)
DISTRIBUTED RANDOMLY;
INSERT INTO array_test_table(content)
VALUES('{8.03,8.033,8.305,8.336,0.838,0.883,8.53,8.586,0.880,0.833,0.883,0.588,0.588,3.885,3.3,3.338,063.858,053.008,80.336,LP,8.655,8.858,8.803,8.608,00.385,88.865,85.088,88.603,85.300,58.086}');

CREATE OR REPLACE FUNCTION array_test()
RETURNS text
VOLATILE
AS $$
# container: plc_python3_shared
import pandas as pd
with plpy.subtransaction():
    df = pd.DataFrame.from_records(
    plpy.execute(
    """
     select * from array_test_table;
    """))
    return "OK"
$$
LANGUAGE plcontainer
EXECUTE ON ANY;
select * from array_test();

```

File: src/sqlhandler.c [hunk 5]
```diff
@@ -310,6 +312,11 @@ plcMessage *handle_sql_message(plcMsgSQL *msg, plcConn *conn, plcProcInfo *pinfo
 						}
 					}
 
+                    /*
```
Comment (beeender) [inline src/sqlhandler.c:315]: indention

File: tests/Makefile [hunk 6]
```diff
@@ -34,7 +34,7 @@ resgroup:
 	plcontainer runtime-show -r plc_r_shared
 
 	PL_TESTDB=$(PL_TESTDB) $(top_builddir)/src/test/regress/pg_regress \
-				$(PSQLDIR) $(RESGROUP_OPTS) || if [[ -f regression.diffs ]]; then cat regression.diffs; exit 1; fi
+				$(PSQLDIR) $(RESGROUP_OPTS)
```
Comment (beeender) [inline tests/Makefile:37]:
```
PostgreSQL regression test driver

Usage:
  pg_regress [OPTION]... [EXTRA-TEST]...

Options:
...
  --print-failure-diffs     Print the diff file to standard out after a failure
```

There is an option for this. Try it

File: src/sqlhandler.c [hunk 7]
```diff
@@ -63,7 +63,14 @@ static plcMsgResult *create_sql_result(bool isSelect) {
 	result->types = palloc(result->cols * sizeof(*result->types));
 	result->names = palloc(result->cols * sizeof(*result->names));
 	result->exception_callback = NULL;
+
+    /*
+    * We do not need free this array because the array iterator will reference it.
```
Comment (beeender) [inline src/sqlhandler.c:68]: indention

File: tests/Makefile (hunk 6 above)
Comment (beeender) [reply tests/Makefile:37]: this is good. Ignore me

**Review by beeender — APPROVED** [review]
:clap:

**Pull request merged**